PsfLauncher/main.cpp

```diff
@@ -125,38 +196,69 @@ int launcher_main(PCWSTR args, int cmdShow) noexcept try
     }
 
-    if (!error.IsThereAnError())
-    {
+    //The only way stopOnScriptError is true is in wait is false.
+    //Which means that if there was an error in the starting script
+    //It was either encountered or the user does not want to stop.
         //Launch underlying application.
-        auto exeName = appConfig->get("executable").as_string().wide();
-        auto exePath = packageRoot / exeName;
-        auto exeArgString = exeArgs ? exeArgs->as_string().wide() : (wchar_t*)L"";
+    auto exeName = appConfig->get("executable").as_string().wide();
+    auto exePath = packageRoot / exeName;
+    auto exeArgString = exeArgs ? exeArgs->as_string().wide() : (wchar_t*)L"";
 
-        //Keep these quotes here.  StartProcess assums there are quptes around the exe file name
-        std::wstring cmdLine = L"\"" + exePath.filename().native() + L"\" " + exeArgString + L" " + args;
-        if (check_suffix_if(exeName, L".exe"_isv))
-        {
-            std::wstring workingDirectory = (packageRoot / dirStr).native();
+    //Keep these quotes here.  StartProcess assums there are quptes around the exe file name
+    std::wstring cmdLine = L"\"" + exePath.filename().native() + L"\" " + exeArgString + L" " + args;
+    if (check_suffix_if(exeName, L".exe"_isv))
+    {
+        std::wstring workingDirectory = (packageRoot / dirStr).native();
 
-            ExecutionInformation execInfo = { exePath.c_str(), cmdLine };
+        ExecutionInformation execInfo = { exePath.c_str(), cmdLine };
 
-            auto currentDirectory = (packageRoot / dirStr);
-            execInfo.CurrentDirectory = currentDirectory.c_str();
-            error = StartProcess(execInfo, cmdShow, false);
-            error.AddExeName(exeName);
-        }
-        else
-        {
-            error = StartWithShellExecute(packageRoot, exeName, exeArgString, dirStr, cmdShow);
-        }
+        execInfo.CurrentDirectory = currentDirectory.c_str();
+        error = StartProcess(execInfo, cmdShow, false);
+        error.AddExeName(exeName);
     }
+    else
+    {
+        error = StartWithShellExecute(packageRoot, exeName, exeArgString, dirStr, cmdShow);
+    }
+
+    //Wait for the starting task to finish.
+    //if we haven't gotten the return value yet.
+    if (startingScriptTask.valid())
+    {
+        error = startingScriptTask.get();
+    }
+
 
     //Launch the end PowerShell script if we are using one.
     auto endScriptInformation = PSFQueryEndScriptInfo();
     if (endScriptInformation)
     {
-        ErrorInformation endingScriptError = RunScript(*endScriptInformation, packageRoot, dirStr, cmdShow);
-        error.AddExeName(L"PowerShell.exe -file ");
+        PowershellScript endingScript(*endScriptInformation, currentDirectory);
+
+        ErrorInformation endingScriptError;
+        bool canScriptRun = false;
+        endingScriptError = CheckIfShouldRun(endingScript.GetRunOnce(), canScriptRun);

```

Comment:
End script runonce is a peculiar requirement. I'm curious on the user story here. I was under the assumption that an end script cleans up what the start script does.

Reply:
True.  That is my intent.  But maybe there is stuff that should be only done once when the application finishes.

At the moment I don't know the user story.  However disabling this feature is trivial and can be removed if need be.  I decided to keep it in here.